File: skills/go-scaffold/scripts/templates/files/internal/users/internal/infrastructure/refresh_token_repository.go
package infrastructure

import (
	"context"
	"errors"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"{{MODULE}}/internal/shared/pgconv"
	"{{MODULE}}/internal/users/internal/application"
	"{{MODULE}}/internal/users/internal/domain"
	"{{MODULE}}/internal/users/internal/infrastructure/db"
)

type RefreshTokenRepository struct {
	q *db.Queries
}

func NewRefreshTokenRepository(pool *pgxpool.Pool) *RefreshTokenRepository {
	return &RefreshTokenRepository{q: db.New(pool)}
}

func toRefreshToken(r db.UsersRefreshToken) domain.RefreshToken {
	return domain.RefreshToken{
		ID:           pgconv.UUIDToString(r.ID),
		UserID:       pgconv.UUIDToString(r.UserID),
		TokenHash:    r.TokenHash,
		FamilyID:     pgconv.UUIDToString(r.FamilyID),
		RevokedAt:    pgconv.TimestamptzToTimePtr(r.RevokedAt),
		ReplacedByID: pgconv.UUIDToStringPtr(r.ReplacedByID),
		ExpiresAt:    pgconv.TimestamptzToTime(r.ExpiresAt),
		CreatedAt:    pgconv.TimestamptzToTime(r.CreatedAt),
	}
}

func (r *RefreshTokenRepository) Create(ctx context.Context, p application.CreateRefreshParams) (domain.RefreshToken, error) {
	row, err := r.q.CreateRefreshToken(ctx, db.CreateRefreshTokenParams{
		ID:        pgconv.StringToUUID(p.ID),
		UserID:    pgconv.StringToUUID(p.UserID),
		TokenHash: p.TokenHash,
		FamilyID:  pgconv.StringToUUID(p.FamilyID),
		ExpiresAt: pgconv.TimeToTimestamptz(p.ExpiresAt),
	})
	if err != nil {
		return domain.RefreshToken{}, err
	}
	return toRefreshToken(row), nil
}

func (r *RefreshTokenRepository) FindByHash(ctx context.Context, hash string) (*domain.RefreshToken, error) {
	row, err := r.q.GetRefreshTokenByHash(ctx, hash)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	t := toRefreshToken(row)
	return &t, nil
}

// Revoke conditionally revokes a still-live token (the query's WHERE revoked_at
// IS NULL). It returns true only when a row was actually updated; false means
// the token was already revoked — the caller (rotation) must then abort rather
// than mint a replacement, which is what closes the refresh TOCTOU race.
func (r *RefreshTokenRepository) Revoke(ctx context.Context, id string, replacedByID *string) (bool, error) {
	rows, err := r.q.RevokeRefreshToken(ctx, db.RevokeRefreshTokenParams{
		ID:           pgconv.StringToUUID(id),
		ReplacedByID: pgconv.StringPtrToUUID(replacedByID),
	})
	if err != nil {
		return false, err
	}
	return rows == 1, nil
}

func (r *RefreshTokenRepository) RevokeFamily(ctx context.Context, familyID string) error {
	return r.q.RevokeRefreshTokenFamily(ctx, pgconv.StringToUUID(familyID))
}

File: skills/nodejs-scaffold/scripts/templates/files/src/modules/users/infrastructure/refresh-token-repository.ts
import { and, eq, isNull } from 'drizzle-orm'
import { db } from '../../../shared/db/client.js'
import { refreshTokens } from './users.schema.js'

type Row = typeof refreshTokens.$inferSelect

export interface CreateRefreshTokenRow {
  userId: string
  tokenHash: string
  familyId: string
  expiresAt: Date
}

export const refreshTokenRepository = {
  async create(input: CreateRefreshTokenRow): Promise<Row> {
    const [row] = await db.insert(refreshTokens).values(input).returning()
    return row
  },

  async findByHash(tokenHash: string): Promise<Row | undefined> {
    const [row] = await db.select().from(refreshTokens).where(eq(refreshTokens.tokenHash, tokenHash))
    return row
  },

  async revoke(id: string, replacedById: string | null): Promise<void> {
    await db.update(refreshTokens).set({ revokedAt: new Date(), replacedById }).where(eq(refreshTokens.id, id))
  },

  // Reuse of an already-revoked token in a family is a theft signal — nuke the
  // whole lineage.
  async revokeFamily(familyId: string): Promise<void> {
    await db
      .update(refreshTokens)
      .set({ revokedAt: new Date() })
      .where(and(eq(refreshTokens.familyId, familyId), isNull(refreshTokens.revokedAt)))
  }
}
